### ISSUE TITLE
ToDataSourceResult and filtering with a decimal or double

### ISSUE THREAD

**Issue by holwerda** [kendo-labs/dlinq-helpers#22]
When I call ToDataSourceResult with and integer filter it works fine; however, when trying to call ToDataSourceResult with a filter of a decimal or double, I am getting this error:

An exception of type 'System.Linq.Dynamic.ParseException' occurred in System.Linq.Dynamic.dll but was not handled in user code.
"Operator '=' incompatible with operand types 'Decimal?' and 'Double'".

![image](https://cloud.githubusercontent.com/assets/1644495/16085826/41f7e94a-32d2-11e6-86b2-de822cb41658.png)

![image](https://cloud.githubusercontent.com/assets/1644495/16085869/618a56e4-32d2-11e6-821c-0ef7ba8d2974.png)

Have you seen this? Any thoughts?

**Comment by patelharshilb** [comment]
Even i am facing the same error when trying with decimal.  filtering only works with string values, apart from that it doesnt work.

Need to find out a solution for that.  for now i have just disabled server side filtering and sorting which is not the correct way, but due to lack of time i had no other option.

I will try to find out solution if i can otherwise will need someones help on this.

**Comment by hiraldesai** [comment]
@holwerda - your issue is not kendo related. You can't do an equal check on nullable Decimal and double. What you're trying to do is equivalent of this 

```
decimal? x;
    double y;

    x = 1;
    y = 1;

    if (x == y)  //<- this will give you compiling error
```

@patelharshilb - filtering works for strings, numbers, booleans. You just need to supply the correct filter.

Although what kendo dynamic linq does is just provide helpers to deal with most common use cases. I have in past had to modify `public string ToExpression(IList<Filter> filters)` method of `Filter.cs` to make it do custom project specific filtering.

I have always wanted to extend it to somehow find the type of the source field and dynamically trying to convert the supplied filter value to that type and creating an appropriate expression. Especially because lot of kendo grids we use in our project have plenty of enum filters on the columns and it'd be nice if the filtering was smart enough to convert string enum value to its enum equivalent before creating the filter expression.

**Comment by holwerda** [comment]
You mention that it should work for numbers.  Everything I specify is decimal. I don't specify double anywhere in my code.  Do I need to explicitly loop through all filters and set the type I want in each filter value to try and force it to use decimal?

**Comment by sarvaratchagan** [comment]
@hiraldesai - Yes, I agree with you.

@holwerda - Could you please post that part of your code here. So that people can Identify your issue and try to help

**Comment by holwerda** [comment]
@sarvaratchagan 

DataSourceRequest has a Filter for SomeDollarAmount:

```
Field: SomeDollarAmount
Filters: null
Logic: "and"
Operator: "eq"
Value: 99818
```

here is a modified version of query(just removed most of the returned columns and renamed the items ):

```
using (var myEntities = new MyEntities())
            {
                var query =
                    from active in myEntities.MyTable
                    join previous in myEntities.MyTable.Where(w => w.SomeColumn == 'columnTextPrevious')
                        on active.AccountId equals previous.AccountId 
                        into p
                    from previous in p.DefaultIfEmpty()
                    where active.SomeColumn == 'columnTextActive'
                    orderby active.SomeOtherColumn descending, active.SomeOtherColumn2
                    select new
                    {                        
                        AccountId= active.AccountId,                       
                        SomeDollarAmount = active.SomeDollarAmount   //defined as decimal in EF code and DB
                    };

                return query.ToDataSourceResult(dataSourceRequest);
          }
```

**Comment by leuwintintellibox** [comment]
FIlter still doesnt work for Decimal values. Anyone here has found any solution to the issue ?

**Comment by sarvaratchagan** [comment]
@holwerda 

There is nothing to do with kendo. Already @hiraldesai  answered for your question that you have to map correct type for your filter. Otherwise you will get issues like above. 

Note : You can use the SQL Server decimal type ([documentation here](https://www.microsoft.com/en-us/download/details.aspx?id=51958)). That allows you to specify a precision and scale for the numbers in the database, which you can set to be the same as in your C# code.

Your question, however, seems to be about mapping the .NET decimal type to the SQL Server decimal type. And unfortunately, that's not straightforward. The SQL Server decimal, at maximum precision, covers the range from -1038+1 to 1038-1. In .NET, however, it can cover anywhere from ±10−28 to ±7.9×1028, depending on the number.

The reason is that, internally, .NET treats decimals very similar to floats—it stores it as a pair of numbers, a mantissa and an exponent. If you have an especially large mantissa, you're going to lose some of the precision. Essentially, .NET allows you to mix numbers which have both a high and a low scale; SQL Server requires you to choose it for a given column ahead of time.

In short, you're going to need to decide what the valid range of decimal values are, for your purposes, and ensure both your program and the database can handle it. You need to do that anyway for your program (if you're mixing values like 10−28 with 7.9×1028, you can't rely on the decimal type anyway).

This is my suggestion. Please correct me if i'm wrong. 
- Thanks

**Comment by holwerda** [comment]
@leuwintintellibox I switched to a different Kendo assembly that is not this one.  It is from the paid edition of Telerik Kendo and seems to be able to handle this situation.   I'm closing this because I have moved on from using this assembly.  But feel free to reopen for your needs.
